PyISY/events.py

```diff
@@ -0,0 +1,299 @@
+        self._writer.flush()
+
+    def connect(self):
+        """Connect to the event stream socket."""
+        if not self._connected:
+            try:
+                self.socket.connect((self.data['addr'], self.data['port']))
+                if self.data.get('tls'):
+                    self.cert = self.socket.getpeercert()
+            except OSError:
+                self.isy.log.error('PyISY could not connect to ISY '
+                                   'event stream.')
+                if self._on_lost_function is not None:
+                    self._on_lost_function()
+                return False
+            self.socket.setblocking(0)
+            self._writer = self.socket.makefile("w")
+            self._connected = True
+            return True
+        return True
+
+    def disconnect(self):
+        """Disconnect from the Event Stream socket."""
+        if self._connected:
+            self.socket.close()
+            self._connected = False
+            self._subscribed = False
+            self._running = False
+
+    def subscribe(self):
+        """Subscribe to the Event Stream."""
+        if not self._subscribed and self._connected:
+            if 'sid' not in self.data:
+                msg = self._mkmsg(strings.sub_msg)
+                self.write(msg)
+            else:
+                msg = self._mkmsg(strings.resub_msg)
+                self.write(msg)
+            self._subscribed = True
+
+    def unsubscribe(self):
+        """Unsubscribe from the Event Stream."""
+        if self._subscribed and self._connected:
+            msg = self._mkmsg(strings.unsub_msg)
+            self.write(msg)
+            self._subscribed = False
+            self.disconnect()
+
+    @property
+    def heartbeat_time(self):
+        """Return the last ISY Heartbeat time."""
+        if self._lasthb is not None:
+            return (datetime.datetime.now() - self._lasthb).seconds
+        return 0.
+
+    def watch(self):
+        """Watch the subscription connection and report if dead."""
+        if self._subscribed:
+            while self._running and self._subscribed:
+                # verify connection is still alive
+                if self.heartbeat_time > self._hbwait:
+                    self.disconnect()
+                    self.isy.log.warning('PyISY lost connection to '
+                                         'the ISY event stream.')
+                    if self._on_lost_function is not None:
+                        self._on_lost_function()
+
+                # poll socket for new data
+                inready, _, _ = select.select([self.socket], [], [], POLL_TIME)
+                if self.socket in inready:
+                    for data in self.read():
+                        if data.startswith('<?xml'):
+                            data = data.strip(). \
+                                        replace('POST reuse HTTP/1.1', '')
+                            self._routemsg(data)
+
+
+class EventEmitter:
```

Comment:
I don't remember where these classes were before, but they probably shouldn't be in the same file as the event stream code. I don't believe they are related.

Reply:
Moved back to the `Nodes/__init__.py` file.  I was trying to reorganize, but you're right, these are only used for nodes.

Reply:
Moved theses into a separate `handlers.py` file to prevent circular import errors (could not import EventEmitter when isy.py imports nodes).